app/services/service_result.rb

```diff
@@ -90,53 +107,74 @@ def initialize(success: false,
     self.dependent_results = dependent_results
   end
 
-  alias success? :success
+  # @see failure?
+  # @return [Boolean] whether the Service call succeeded.
+  alias success? success
 
+  # @see success?
+  # @return [Boolean] whether the Service call failed.
   def failure?
     !success?
   end
 
   ##
-  # Merge another service result into this instance
-  # allowing optionally to skip updating its service
+  # Merge another ServiceResult into this instance, optionally allowing its {#success} to be ignored.
+  #
+  # @param other [ServiceResult] the other ServiceResult.
+  # @param without_success [Boolean] whether to ignore the {#success} of the other ServiceResult.
+  # @return [void]
   def merge!(other, without_success: false)
     merge_success!(other) unless without_success
     merge_errors!(other)
     merge_dependent!(other)
   end
 
   ##
-  # Print messages to flash
+  # Print messages to the Controller's flash.
+  #
+  # @param flash [ActionDispatch::Flash::FlashHash]
+  # @return [void]
   def apply_flash_message!(flash)
     if message
       flash[message_type] = message
     end
   end
 
+  ##
+  # Returns all {#result}s, including from dependent ServiceResults.
+  #
+  # @return [Array] all results.
```

Comment:
I think you can note this as `Array<Object, nil>` to imply we have no idea what the results can be.

Reply:
In this case, `all_results` filters out any `nil` results, so this will be `Array<Object>`.

Reply:
BTW, I wonder if we could change the implementation to `filter_map` 🤔

Reply:
@mereghost I was thinking the same thing:

```rb
    self_and_dependent.flat_map(&:result)
```

However it would introduce a behaviour change, as the current implementation can contain nil results from dependent ServiceResults. (it might be that that wasn't the intention, but since I am not the original author that is only me speculating)